### DISCUSSION TITLE
Best practices and optimizations of backends usage

### DISCUSSION THREAD

**Discussion by NBardelot** [apache/airflow#38073] [Q&A]
# Some context and references

* [#30265 Adding a cache to Variable.get](https://github.com/apache/airflow/discussions/30265) 
* [The #30265 follow up on the Apache mailing list](https://lists.apache.org/thread/qmof64dtoc24nh1r1jsq23flbn5jrzbq)
* [Astronomer's guidelines for storing information in Airflow](https://docs.astronomer.io/learn/airflow-variables#best-practices-for-storing-information-in-airflow)
* [Airflow's best practice for DAGs top-level code](https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#top-level-python-code)
* [Raphaël Vandon's blog entry about the caching of secrets introduced in Airflow 2.7](https://medium.com/apache-airflow/the-ins-and-outs-of-airflows-new-secrets-cache-f7b9ec25ca1e)

My organization uses multiple Airflow instances, that we manage on AWS EKS + Hashicorp Vault, with hundreds of DAGs per instance. Development is decentralized (~50 devs, not specialized in writing Airflow code), and my team provides Airflow instances as a PaaS to the IT teams. We're writing our guidelines.

We face a similar issue than other Airflow users concerning the best practices for storing and retrieving Airflow Variables. Here are some of our current guidelines based on the links mentioned earlier:

* minimize calls to `Variable.get` in top-level code
* storing only sensitive data in our Hashicorp Vault; leave the non-sensitive data in the metadata database (RDS/PostgreSQL in our case)
* activate caching with a 15 min TTL (using the feature introduced in Airflow 2.7)
* use os.environ as much as possible for *static* parametrization (for exemple the AWS account ID, or the name of the instance... things that can be configured as env vars in the container at deployment time)
* merge multiple variables in one JSON variable when possible

Yet, one use-case we often see in our DAGs is that people want to parametrize the "schedule" of their DAGs, in order to run the DAG in different modes depending on manually configured cron rules that change over time. And even one variable like that at top-level, because it is multiplied by hundreds of DAGs, is leading to some issues tied to how Airflow accesses the Vault/Metadata DB. We cannot simply ask people to push a git commit to change the schedule, since it is then considered as a code change, and needs a change validation (our ITIL process).

We've identified several questions for the Airflow maintainers that we'd like to discuss in order to better understand what is at stake in the Airflow components; how we can mitigate the effets; and what would be the future of Airflow's best practices depending on the architectural choices the maintainers have made and/or will make.

1. Cache predecence
2. Behaviour of the DAG processor regarding Variables (and possible optimizations)
3. Misses and retries for the Hashicorp Vault backend
4. DAGs parametrization (especially "schedule")

# Cache precedence

You can see on the diagram on Raphaël Vandon's / @vandonr-amz blog entry, the current behaviour:

![Diagram of precedence](https://miro.medium.com/v2/resize:fit:720/format:webp/0*BY3-ge88pS2IrzpA)

With this precedence, if a data is not sensitive, and stored in the Metadata DB, the secrets backend is called nevertheless (with a 404-like response). Wouldn't it be more effective to put the secrets backend after the Metadata DB? Thus, you'd put pressure on the secrets backend only for secrets, leaving the Metadata DB responsible for non-sensitive data, manageable in the Airflow UI. 

As such a change would be a breaking change, this could be implemented with a "precedence" feature switch in the parameters of the secrets backend.

By the way, Variable.set does store the data in the Metadata DB and not in the secrets backend, which is counter-intuitive for the current precedence (get/set do not behave the same way).

# Behaviour of the DAG processor regarding Variables

Along discussions between the maintainers and power-users on the Apache mailing list, and accross documentation, we see that @potiuk mentions some "housekeeping" is done in the DAGFileProcessor. But I could not find any detailed explanation. Could you tell us what is done during DAGFileProcessor that is not dag parsing?

The Astronomer documentation mentions a workaround, which consists in using Jinja2 templates instead of `Variable.get` in order to avoid real network access to the variable. I suppose this means that the templating engine is not running for the DAGFileProcessor, and only when the task instance is really executed. Am I right?

If my previous assumption is right, we have the idea of wrapping `Variable.get` with `if` statements based on `os.environ` variables. We'd call `Variable.get` only if the container running the top-level code is **not a DAG processor**. Would it work? Are there cases where the DAG processor really needs to know the real value of a Variable?

# Misses and retries for the Hashicorp Vault backend

We put quotas in place for our secret engines. Two cases have to make a burst of calls to the backend:

* starting an Airflow instance (getting all the configuration for each Airflow component)
* DAG processing

While the second one is debatable (good practice should mitigate the issue), at least for the first one there is room for improvement. The secrets backend classes could support delaying and retrying when quota is exceeded, by implementing a backoff with jitter pattern.

# DAGs parametrization

A very quick-win regarding DAG constructors params seems to be able to provide a second constructor to the DAG class, which would take a static callback function instead of static params. This way, we could apply the same best practice that is documented for other `Variable.get` calls, and put the code that does the heavy-loading apart from the top-level code.

If calling a static function in a constructor is considered not good enough (I'm not python expert...), maybe a factory pattern where the constuction is delegated could be implemted.

--

We're curious to known your thoughts on those topics, and to get feedback both from the maintainers and the community. 

Best regards, and thanks for coding Airflow.

**Comment by potiuk** [comment]
You should take a look at https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#use-cache added in 2.7.0. it should directly address your issue.

This is an experimental feature - but [leaase try it out and give us a feedback. Having good feedback from customer like you who experienced problems and (hopefully) will get their problem addressed might help us to decide to move it out experimental.

**Comment by NBardelot** [comment]
Hi and thanks @potiuk. We already use the feature. It alleviates the issue a great deal, especially when we start Airflow and retrieve the configuration from the secrets backend, but it's not yet enough in our case.

**Reply by potiuk** [reply]
You should then employ Vault caching agent.

**Reply by potiuk** [reply]
BTW.  You can of course propose changing the precedence - there were a few discussions and PRs around but so far no-one proposed a good solution and followed it through. If you think you can add PR. Any other proposals you had there is mostly a hack.

**Reply by potiuk** [reply]
> By the way, Variable.set does store the data in the Metadata DB and not in the secrets backend, which is counter-intuitive for the current precedence (get/set do not behave the same way).

It's been better documented  in a recent PR.

**Reply by potiuk** [reply]
>  but it's not yet enough in our case

Not sure what's enough / not enough. But if you could elaborate what "not enough" means, it would be helpful

**Reply by potiuk** [reply]
> Yet, one use-case we often see in our DAGs is that people want to parametrize the "schedule" of their DAGs, in order to run the DAG in different modes depending on manually configured cron rules that change over time. 

This is an extremely bad pattern to change schedule dynamically. You should avoid that. DAGs should have mostly static schedule - and if you have similar DAGs with different schedule, you should use Dynamic DAGs to generate DAGs with different schedule.

**Reply by potiuk** [reply]
> Yet, one use-case we often see in our DAGs is that people want to parametrize the "schedule" of their DAGs, in order to run the DAG in different modes depending on manually configured cron rules that change over time.

If you REALLY want to do it (and change it dynamically) - add a process where chage from your users is propagated to the environment variable and you restart DAG file processor after that taking that environment variable. You are generally abusing Airflow Variables for things they were never intended to do, and with such custom requirement, implementig it in a custom way and making your deployment a bit more complex rather than complicating the product for such a niche case is way better solution.

**Reply by NBardelot** [reply]
> You should then employ Vault caching agent.

We're looking into it, with a Vault proxy sidecar in cache mode (in addition to the Airflow cache itself). The issue is that we're not sure if the Vault proxy can cache 404s. If it cannot, then with the current precedence a new Variable (or a Variable that has reached its TTL) will be accessed in the Vault anyway. Only "found" secrets will correctly be cached. 

It would look like this for example:

1. Airflow asks the cache if the data is here? No it is not; or yes but with a TTL reached...
2. Airflow asks the proxy sidecar if the data is here? No it is not, since it is not a secret but a non-sensitive data. -> and we don't know if this 404 is cached... if it is not, then the sidecar will propagate the request to the backend
3. Airflow ends up asking the Metastore if the data is here? Yes it is, and it is cached.

As you can see, even with a proxy sidecar the precedence can make the backend being accessed a lot.

I'll follow up on this when we'll be sure what the behaviour is.

> Not sure what's enough / not enough. But if you could elaborate what "not enough" means, it would be helpful

Vault is a critical component of our infrastructure, and we cannot avoid setting quotas. Usual quotas are around a few tens of requests per second depending on the engine. In our case we see bursts of requests to the secret engine that go beyond a hundred calls per second, and the bursts increase proportionnaly with  the number of DAGs, which means the architecture isn't scalable as-is. 

Plus, if we're not careful enough, teams can add DAGs and exceed the quota, which ends up creating issues in production. Though this could be avoided if the secrets backend had a backoff/jitter behaviour (which itself is not a "hack" but a usual networking/protocol resilience pattern).

> This is an extremely bad pattern to change schedule dynamically. You should avoid that. DAGs should have mostly static schedule - and if you have similar DAGs with different schedule, you should use Dynamic DAGs to generate DAGs with different schedule.

Changing "when" a DAG is ran does not make it another DAG, from a functional standpoint (in our case, a business process). If the scheduling could be expressed with a cron rule, it would be done, but some scheduling patterns cannot be expressed as only one cron rule and are more complex than that.

Plus, from an ITIL standpoint, changing "when" a DAG is ran is a configuration change, not a code change. And we treat pushing code differently than configuring a variable in the Airflow UI. Our teams have legitimate use cases for modifying a DAG schedule depending on human business inputs.

That being said, do you think we can use dynamic DAGs with a "schedule" as a Jinja2 template? This way, we could read from the backends only when needed.

**Reply by potiuk** [reply]
Generally Speaking DON't Use Vault for schedule. This is a dead end. Build another mechanism for that.

**Reply by potiuk** [reply]
BTW. This is really an ultimate answer, reading schedule from Variable is NOT and it is unlikely it will ever be a feature of Airflow. We are not likey going to complicate airflow architecture for your particular case, and it's generally abuse of Vault and secrets that they were never intended to handle. 

My feeling is that You will loose time (of yours and others) trying to change it for "Airflow" as a product and this time (of yours especially) can be spent better by developing a custom mechanism - for example with env variables that I suggested you. You could proabbly even write a little nice util for your deployment where you will be able to provide a mechanism to update schedule via env variables or some other ways - but I can't see how such mechanism will ever make it into a product feature.

What I **may** propose you as the best course of action:

* develop custom solution for your case
* perfect it and test
* when you have it ready and think you can generalize it, come here with AIP document decribing your proposal and discuss it on the devlist
*  during that ask others if they feel that woudl be a feature they would like to use, build consensus, gain support. Devlist is the right place, and having a working implementation in your deployment that you can show that it works and that it is useful is a big plus.

You are probably the most interested person in the world (and about one of very few) who wants this kind of solution, so it only makes sense that you implement and prove it in your own deployment and bring it as a feature when you have some proven results of it.

**Reply by NBardelot** [reply]
Just to be clear, I'm not saying I want any technical feature at all cost. On the contrary, I'm trying to assert what is the best way to reach the desired business behaviour.

I'll take a look at [Implementing Event-Based DAGs with Airflow
](https://airflowsummit.org/sessions/2022/implementing-event-based-dags-with-airflow/) (Airflow summit 2022). Maybe a form of event-driven DAG is the way for what our users want.

**Reply by potiuk** [reply]
Understd. But really abusing variables to do dynamic schedule is bad idea. What you are after is to decouple schedule from dag structure and task code - which is precisely what airflow does not do today - all that is coupled in the DAG

**Reply by potiuk** [reply]
So you are after a big, fundamental change in Airflow where you want to decouple schedule definition from dag structure code and execution code. This is what you are after. And yes  it might be a proposal to make but again the best way to do it is that you spend your resources and time to try it out, see it and propose it back once you get something like that worki g and you are reasonably sure it can be generalized enough to turn into a product feature. This is like most of the features I. airflow are created

**Reply by potiuk** [reply]
And having co panyblike you try it and develop and then lead and propose something like that after proving it's value, showing the use case, gaining consensus in a discusion, getting approvals is a great idea and fantastic way to give back to the community for the free software you use

**Reply by potiuk** [reply]
So I am not discouraging you, just telling that today's architecture of airflow is not suited to your case - but with sufficient investment of your team where you try and proof what you think might be done, you might turn it into a feature - since You need it makes sense you implement it and then with learnings bring it to the community as a proposal - that's how often things work here

**Comment by NBardelot** [comment]
FYI, a colleague has spotted that [Hashicorp will enable static secret caching](https://vault-r1xrfcpnr-hashicorp.vercel.app/vault/docs/agent-and-proxy/proxy/caching/static-secret-caching) soon (it's a feature in their roadmap).

It will allow to also cache misses, which will alleviate some of the load due to the current precedence in the way Airflow's hashicorp Vault secret backend is implemented.

> When static secret caching is enabled, Proxy returns HIT or MISS in the X-Cache response header for requests so client can tell if the response was served from the cache or forwarded from Vault. In the event of a hit, Proxy also sets the Age header to indicate, in seconds, how old the cache entry is.

(See also: [issue 19879 on Hashicorp Vault's project](https://github.com/hashicorp/vault/issues/19879#issuecomment-1830075509))
